### ISSUE TITLE
Question: How to pass function?

### ISSUE THREAD

**Issue by kanatapple** [madsmtm/objc2#269]
Hi, How can I pass a function to msg_send and call it?
For example, I want to pass a function to the completionHandler of setCookie method in WKHTTPCookieStore like in the code below.

```
// https://developer.apple.com/documentation/webkit/wkhttpcookiestore/2882007-setcookie?language=objc
msg_send![http_cookie_store, cookie: cookie_data, completionHandler: function];
```
I don't know how to define function.

I'm actually using objc, but this community is more active so I asked here.

**Comment by madsmtm** [comment]
The `completionHandler` argument is _not_ a function, rather, it is a "block" (see [this](https://clang.llvm.org/docs/BlockLanguageSpec.html) for details, essentially the C-equivalent of a Rust `dyn Fn` closure).

So to call it, you'd do something like this:
```rust
use block2::{Block, ConcreteBlock}; // or the old `block` crate

let handler = ConcreteBlock::new(|| {
    // Code to run on completion.
    // This closure can capture values, but since `completionHandler` is executed asynchronously, the captured
    // values must be `Send + Sync`.
});
// Put the block on the heap
let handler = handler.copy();
// Ensure that the block has the correct type (no arguments, `void` return type)
let handler: &Block<(), ()> = &handler;

let _: () = unsafe { msg_send![http_cookie_store, cookie: cookie_data, completionHandler: handler] };
```

**Comment by madsmtm** [comment]
(I know that the API that the `block2` crate exposes is not really that nice to work with, I am working on improving this, see #168)

**Comment by kanatapple** [comment]
Thanks. I will try.

**Comment by kanatapple** [comment]
It's working as expected.
Thank you.

**Comment by silvanshade** [comment]
@madsmtm Is the example above still how one is supposed to pass a block as an argument? It doesn't seem to work with the published crates (`block2 = "0.1.6"` and `icrate = "0.0.1"`).

With the following simplified example, I get an error:

```
use block2::{Block, ConcreteBlock};
use icrate::{objc2::*, Foundation::NSObject};

fn main() {
    let block = ConcreteBlock::new(|| {});
    let block = block.copy();
    let block: &Block<(), ()> = &block;
    let object = NSObject::new();
    unsafe { msg_send![&object, completionHandler: block] }
}
```

with the message:

```
the trait bound `block2::Block<(), ()>: RefEncode` is not satisfied
the following other types implement trait `RefEncode`:
  &'a T
  &'a mut T
  *const T
  *const c_void
  *mut T
  *mut c_void
  AtomicI16
  AtomicI32
and 911 others
required for `&block2::Block<(), ()>` to implement `Encode`
required for `&block2::Block<(), ()>` to implement `EncodeConvert`
required for `(&block2::Block<(), ()>,)` to implement `MessageArguments`rustc[Click for full compiler diagnostic]
```
